flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaFetcher.java

```diff
@@ -0,0 +1,238 @@
+			processingTimeProvider,
+			autoWatermarkInterval,
+			userCodeClassLoader,
+			consumerMetricGroup,
+			useMetrics);
+
+		this.deserializer = deserializer;
+		this.handover = new Handover();
+
+		this.consumerThread = new KafkaConsumerThread(
+			LOG,
+			handover,
+			kafkaProperties,
+			unassignedPartitionsQueue,
+			createCallBridge(),
+			getFetcherName() + " for " + taskNameWithSubtasks,
+			pollTimeout,
+			useMetrics,
+			consumerMetricGroup,
+			subtaskMetricGroup);
+	}
+
+	// ------------------------------------------------------------------------
+	//  Fetcher work methods
+	// ------------------------------------------------------------------------
+
+	@Override
+	public void runFetchLoop() throws Exception {
+		try {
+			final Handover handover = this.handover;
+
+			// kick off the actual Kafka consumer
+			consumerThread.start();
+
+			while (running) {
+				// this blocks until we get the next records
+				// it automatically re-throws exceptions encountered in the consumer thread
+				final ConsumerRecords<byte[], byte[]> records = handover.pollNext();
+
+				// get the records for each topic partition
+				for (KafkaTopicPartitionState<TopicPartition> partition : subscribedPartitionStates()) {
+
+					List<ConsumerRecord<byte[], byte[]>> partitionRecords =
+						records.records(partition.getKafkaPartitionHandle());
+
+					for (ConsumerRecord<byte[], byte[]> record : partitionRecords) {
+						final T value = deserializer.deserialize(
+							record.key(), record.value(),
+							record.topic(), record.partition(), record.offset());
+
+						if (deserializer.isEndOfStream(value)) {
+							// end of stream signaled
+							running = false;
+							break;
+						}
+
+						// emit the actual record. this also updates offset state atomically
+						// and deals with timestamps and watermark generation
+						emitRecord(value, partition, record.offset(), record);
+					}
+				}
+			}
+		}
+		finally {
+			// this signals the consumer thread that no more work is to be done
+			consumerThread.shutdown();
+		}
+
+		// on a clean exit, wait for the runner thread
+		try {
+			consumerThread.join();
+		}
+		catch (InterruptedException e) {
+			// may be the result of a wake-up interruption after an exception.
+			// we ignore this here and only restore the interruption state
+			Thread.currentThread().interrupt();
+		}
+	}
+
+	@Override
+	public void cancel() {
+		// flag the main thread to exit. A thread interrupt will come anyways.
+		running = false;
+		handover.close();
+		consumerThread.shutdown();
+	}
+
+	protected void emitRecord(
+		T record,
+		KafkaTopicPartitionState<TopicPartition> partition,
+		long offset,
+		ConsumerRecord<?, ?> consumerRecord) throws Exception {
+
+		emitRecordWithTimestamp(record, partition, offset, consumerRecord.timestamp());
+	}
+
+	/**
+	 * Gets the name of this fetcher, for thread naming and logging purposes.
+	 */
+	protected String getFetcherName() {
+		return "Kafka 2.0 Fetcher";
```

Comment:
Probably don't want to be updating this version every time there is a new Kafka release.

Reply:
Can we give it a unified name? What about `Kafka Fetcher`?